Update beeper 4.0.623 and add zh4ngx as maintainer

zh4ngx opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

zh4ngx commented:
Fixes #389967

zh4ngx commented:
@ymstnt Thanks for the help!

dtomvan commented:
You should split this into two commits:
```
beeper: 3.110.1 -> 4.0.584
beeper: add zh4ngx as maintainer
```

dtomvan commented:
Builds and runs correctly :+1:.

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 395624`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>beeper</li>
  </ul>
</details>

jiriks74 commented:
As @dtomvan mentioned - split the commits into two.

Also read [Contributing.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md) where there are commit message convetions. Specifically [this section](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions) for `pkgs`. Then check that it `Fits CONTRIBUTING.md.`:

> Commit your changes using `git commit`. Make sure to adhere to the [commit conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions).

zh4ngx committed:
beeper: 3.110.1 -> 4.0.623

zh4ngx committed:
beeper: add zh4ngx as maintainer

zh4ngx commented:
Thanks for the review and the tips. Read through the doc and updated the commits.

jiriks74 reviewed (approved):
LGTM

jshcmpbll reviewed (approved):
LGTM!

If you are interested here are the changes to improve the update script. I can add them in a separate PR too if you would prefer.

Running the update script from master with changes:

```bash
nixos/nixpkgs nix-shell maintainers/scripts/update.nix --argstr package beeper
these 2 derivations will be built:
  /nix/store/p68v7ax4gy2vc5kl9q5w9ikdjwwnw3ly-update-beeper.drv
  /nix/store/yrjjgq8k988jc6zl4w5z19j0chsd2rwp-packages.json.drv
building '/nix/store/p68v7ax4gy2vc5kl9q5w9ikdjwwnw3ly-update-beeper.drv'...
building '/nix/store/yrjjgq8k988jc6zl4w5z19j0chsd2rwp-packages.json.drv'...

Going to be running update for following packages:
 - beeper-3.110.1

Press Enter key to continue...

Running update for:
Enqueuing group of 1 packages
 - beeper-3.110.1: UPDATING ...
 - beeper-3.110.1: DONE.

Packages updated!
```

Git changes:

```bash
diff --git a/pkgs/by-name/be/beeper/package.nix b/pkgs/by-name/be/beeper/package.nix
index e25728089d55..cc6655c1558a 100644
--- a/pkgs/by-name/be/beeper/package.nix
+++ b/pkgs/by-name/be/beeper/package.nix
@@ -11,10 +11,10 @@
 }:
 let
   pname = "beeper";
-  version = "3.110.1";
+  version = "4.0.584";
   src = fetchurl {
-    url = "https://download.beeper.com/versions/3.110.1/linux/appImage/x64";
-    hash = "sha256-AwPYwA93zOxExxhkLpCUKsQC/ylKSAygdHBUjd5ZKZk=";
+    url = "https://beeper-desktop.download.beeper.com/builds/Beeper-4.0.584.AppImage";
+    hash = "sha256-xKOVU1AiOfC7QIMn0ErUDVOVjho6cm2qXY1X2zk2kgs=";
   };
   appimage = appimageTools.wrapType2 {
     inherit version pname src;
@@ -56,14 +56,13 @@ stdenvNoCC.mkDerivation rec {
       name = "update-beeper";
       runtimeInputs = [
         curl
-        yq
         common-updater-scripts
       ];
       text = ''
         set -o errexit
-        latestLinux="$(curl -s https://download.todesktop.com/2003241lzgn20jd/latest-linux.yml)"
-        version="$(echo "$latestLinux" | yq -r .version)"
-        update-source-version beeper "$version" "" "https://download.beeper.com/versions/$version/linux/appImage/x64" --source-key=src.src
+        latestLinux="$(curl --silent --output /dev/null --write-out "%{redirect_url}\n" https://api.beeper.com/desktop/download/linux/x64/stable/com.automattic.beeper.desktop)"
+        version="$(echo "$latestLinux" |  grep --only-matching --extended-regexp '[0-9]+\.[0-9]+\.[0-9]+')"
+        update-source-version beeper "$version" "" "https://beeper-desktop.download.beeper.com/builds/Beeper-$version.AppImage" --source-key=src.src
       '';
     });
   };
```

zh4ngx commented:
Thanks! Should be all set

zh4ngx commented:
> @zh4ngx if you run this at the root of the repo in your working branch it should automatically fetch the new version and update the hash for you.
> 
> ```
> nix-shell maintainers/scripts/update.nix --argstr package beeper
> ```

I'm getting an error when I try to run the script:

```bash
nix-shell maintainers/scripts/update.nix --argstr package beeper


Going to be running update for following packages:
 - beeper-4.0.584

Press Enter key to continue...
Running update for:
Enqueuing group of 1 packages
 - beeper-4.0.584: UPDATING ...
 - beeper-4.0.584: ERROR

--- SHOWING ERROR LOG FOR beeper-4.0.584 ----------------------

error: attribute 'src' missing
       at «string»:1:142:
            1| let pkgs = (let tree = import ./.; in if builtins.isFunction tree then tree {} else tree); in pkgs.beeper.src.src.drvAttrs.outputHashAlgo or pkgs.beeper.src.src.drvAttrs.outputHash
             |                                                                                                                                              ^
error: attribute 'src' in selection path 'beeper.src.src.drvAttrs.outputHash' not found
update-source-version: error: Couldn't evaluate old source hash from 'beeper.src.src'!


--- SHOWING ERROR LOG FOR beeper-4.0.584 ----------------------
The update script for beeper-4.0.584 failed with exit code 1
```

Do you think it has something to do with switching to `appimageTools.wrapAppImage`?

zh4ngx commented:
@jshcmpbll just updated and rebased

zh4ngx commented:
@jshcmpbll just updated and rebased

> > @zh4ngx if you run this at the root of the repo in your working branch it should automatically fetch the new version and update the hash for you.
> > ```
> > nix-shell maintainers/scripts/update.nix --argstr package beeper
> > ```
> 
> I'm getting an error when I try to run the script:
> 
> ```shell
> nix-shell maintainers/scripts/update.nix --argstr package beeper
> 
> 
> Going to be running update for following packages:
>  - beeper-4.0.584
> 
> Press Enter key to continue...
> Running update for:
> Enqueuing group of 1 packages
>  - beeper-4.0.584: UPDATING ...
>  - beeper-4.0.584: ERROR
> 
> --- SHOWING ERROR LOG FOR beeper-4.0.584 ----------------------
> 
> error: attribute 'src' missing
>        at «string»:1:142:
>             1| let pkgs = (let tree = import ./.; in if builtins.isFunction tree then tree {} else tree); in pkgs.beeper.src.src.drvAttrs.outputHashAlgo or pkgs.beeper.src.src.drvAttrs.outputHash
>              |                                                                                                                                              ^
> error: attribute 'src' in selection path 'beeper.src.src.drvAttrs.outputHash' not found
> update-source-version: error: Couldn't evaluate old source hash from 'beeper.src.src'!
> 
> 
> --- SHOWING ERROR LOG FOR beeper-4.0.584 ----------------------
> The update script for beeper-4.0.584 failed with exit code 1
> ```
> 
> Do you think it has something to do with switching to `appimageTools.wrapAppImage`?

Fixed the issue by following the appimageTools guide and switching to `wrapType2`.

zh4ngx commented on pkgs/by-name/be/beeper/package.nix:
```diff
@@ -1,69 +1,57 @@
-        update-source-version beeper "$version" "" "https://download.beeper.com/versions/$version/linux/appImage/x64" --source-key=src.src
+        latestLinux="$(curl --silent --output /dev/null --write-out "%{redirect_url}\n" https://api.beeper.com/desktop/download/linux/x64/stable/com.automattic.beeper.desktop)"
+        version="$(echo "$latestLinux" |  grep --only-matching --extended-regexp '[0-9]+\.[0-9]+\.[0-9]+')"
+        update-source-version beeper "$version" "" "https://beeper-desktop.download.beeper.com/builds/Beeper-$version.AppImage"
```
Great catch! Indeed, it's now behaving as expected.

zh4ngx replied:
Sorry, still figuring out how to make this `updateScript` work with the rest of the build. The issue is we either need to use `wrapAppImage` in order to bring over the modified `registerLinuxConfig()` function from `linux-*.mjs` file, or we need another way to modify that file in the final output such that `registerLinuxConfig()` is a no-op without having to use `wrapAppImage`. 

As an aside, I logged a bug with Beeper, hoping they will either wipe out this function or at least move it into a button or cmd-line option. Per [this comment here](https://github.com/NixOS/nixpkgs/issues/389967#issuecomment-2774596232), that js script is not doing anything correct/useful. I confirmed after inspecting the code that it was inconsistent with the provided desktop file and contained hardcoded paths.

Are you familiar with `wrapAppImage` and how it behaves differently from `wrapType2`? If we can fix the `update-source-version` to adjust accordingly, this will solve our problem until the `postExtract` is no longer needed - we can easily move the `substituteInPlace` call under `wrap*`.`extraInstallCommands`

In case you want to give it a try as well, I tried basing off of the original structure from `master` where we have `extract`, `wrap`, and `mkDerivation`, but it seemed like coping the `resources` folder had no effect on the offending `linux-*.mjs` script, which is probably being referenced elsewhere.

zh4ngx replied:
After digging into the code, I have a clean solution that might require some documentation. 

Rather than hacking in / duplicating the `wrapType2` code (in particular the `src` passthru) to use with `wrapAppImage`, it makes more sense to update `wrapType2` to pass through the `postExtract` feature created in #261190.
I noticed other PRs that ran into the same issue, such as #295447, and it seems like the intended use of `wrapType2` is to run `wrapAppImage` on the `src` of `extract`, which `postExtract` should apply.

I'll update this PR accordingly and create a new one for `postExtract`. I did also review other uses of `postExtract` and confirmed this PR would be the first one to use it in `wrapType2`, so there should be no behavior change to other packages. However, I do think it makes sense to document this change; furthermore, in instances where you don't need to reference `appimageContents` (for example, if only a `substituteInPlace` is needed), you could just use `wrapType2` with `postExtract` and greatly simplify.

RossComputerGuy reviewed (approved):
![i approb](https://github.com/user-attachments/assets/927201e1-43b5-40bf-9272-011afa823fe7)

zh4ngx commented:
One thing to note is this depends on #399356 - otherwise `postExtract` won't run.
